proposals/269/transport.md

```diff
@@ -0,0 +1,74 @@
+---
+Status: IN PROGRESS  
+Author: Julien Le Dem (original issue) / Maciej Obuchowski (this document)  
+Created: 2021-10-18  
+Issue: [269](https://github.com/OpenLineage/OpenLineage/issues/269)
+---
+
+##Purpose
+
+The goal is to facilitate sending `OpenLineage` events over various protocols. 
+(HTTP, queues, ...) independently of the logic via exposing `Transport` interface.
+
+## Proposed interface
+
+In pseudocode:
+```
+interface Transport {
+  send(LineageEvent event);
+}
+```
+
+The send method should be synchronous and return when underlying system confirms 
+that message has been received. This can be for example HTTP response in 200 range, 
+successful fsync on file descriptor, or successful Apache Kafka flush.
+
+Potential error handling is deliberately unspecified on overall interface level, 
+but should be specified and consistent for particular language implementation. 
+The reason for that is difference between idiomatic error handling between different 
+languages: while Python and Java prefer exceptions, Go prefers multiple return values, 
+and Rust prefers monadic `Result<T,E>`. 
+
+However, `Clients` and `Transports` utilizing exceptions should be careful to not directly expose 
+integrations to exceptions thrown by underlying libraries - integrations should be 
+independent of `Transports`. 
+
+
+`Transports` are implemented for particular language's `Client`. There is no guarantee 
+of availability of any `Transport` on instance that implementation exist for different 
+language, with following exception.
+Any reference implementation of `OpenLineage` `Client` should include default 
+`HTTP Transport`. 
+
+## Instantiation
+
+Integration code instantiation is usually controlled by system for which they are 
+written - for example Airflow spawns `LineageBackend` instance. Then, integration
+code would spawn `Client`, which would be responsible for spawning relevant 
+`Transport`.  
+This is contrary to usual dependency injection approach, and creates two issues:   
+that `Client` needs to be aware of which `Transport` to spawn;
+and that `Transport` instantiation should be lightweight and not depend on variables
+passed to it's constructor.  
+Former issue can be solved by standarizing `OPENLINEAGE_TRANSPORT` environment 
+variable notifying `Client` of which `Transport` to spawn.  
```

Comment:
I think we can go ahead and define a `TransportFactory` interface or something that is responsible for determining which client to instantiate. Can we flesh out some details here? Are users to supply the fully qualified class name of the transport (e.g., `OPENLINEAGE_TRANSPORT=io.openlineage.transport.KafkaTransport`)? Will the class names need to be the same across languages or is it ok to differ? or should we use simpler aliases (e.g., `OPENLINEAGE_TRANSPORT=kafka`) and let the factory figure out the correct class to instantiate?

Reply:
Updated it again.